hid.1
.\" Text automatically generated by txt2man
.TH hid  "10 December 2014" "" ""
.SH NAME
\fBhid \fP- interact with attached HID-compliant devices.
.SH SYNOPSIS
.nf
.fam C
\fBhid\fP [\fB--help\fP] <command> [\fIargs\fP]
.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
\fBhid\fP is a tool to send commands to an attached HID-compliant device. 
.PP
\fBhid\fP can also list all attached USB devices.
.SH BASIC COMMANDS
.TP
.B
list
List attached HID devices
.TP
.B
info
List information on attached \fBhid\fP devices
.TP
.B
getreport
Get a report from an attached device
.TP
.B
setreport
Set a report for an attached device
.TP
.B
attached
Check whether a matching device is attached or not
.SH GENERAL OPTIONS
.TP
.B
\fB-n\fP <int>
Run the given command <int> times [default: 1]
.SH FILTERING DEVICES

Arguments can be provided to filter the devices to run the command against.
example: \fBhid\fP list \fB-v\fP 0x19b5
.PP
.nf
.fam C
    -l <hex>   Filter by the given location ID
    -v <hex>   Filter by the given vendor ID
    -p <hex>   Filter by the given product ID

.fam T
.fi
See '\fBhid\fP help <command>' for more information on a specific command.
.SH COMMANDS 

.TP
.B
list
[\fB-vlp\fP arg]
.PP
.nf
.fam C
      List attached HID-compliant devices

.fam T
.fi
.TP
.B
info
[\fB-vlp\fP arg]
.PP
.nf
.fam C
      View information about attached HID-compliant devices.

.fam T
.fi
.TP
.B
getreport
[\fB-t\fP input|output|feature] [\fB-i\fP <report_id>] [\fB-r\fP] [\fB-o\fP] [\fB-x\fP]
.PP
.nf
.fam C
      Get a report from an attached HID-compliant device.

      If more than one device is matched, the first-discovered matching device is used.

      -t <type>  One of 'input', 'output' or 'feature' [default: input]
      -i <id>    Use the given report ID [default: 0]
      -r         Do not remove the report ID byte from the retreived report
      -o         Do not add a newline to the end of the retreived report
      -x         Output the report as human readable hex

.fam T
.fi
.TP
.B
setreport
[\fB-t\fP output|feature] [\fB-i\fP <id>] [\fB-w\fP <time>] [\fB-r\fP] [\fB-o\fP] [\fB-x\fP] \fB-d\fP <data> [data\.\.\.]
.PP
.nf
.fam C
      Send a report to an attached HID-compliant device.

      If more than one device is matched, the first-discovered matching device is used.

      -d <data>  The data of the report, as hex (prefix with 0x) or as a string.
                 Multiple data sets can be passed, space separated, and will be set in order.
                 Report data can also be piped over STDIN, multiple reports are not supported
                 in this case
      -t <type>  One of 'output' or 'feature' [default: output]
      -i <id>    Use the given report ID [default: 0]
      -w <time>  If provided, the command will wait for an input report for the
                 given number of seconds
      -r         Do not remove the report ID byte from any received input reports
      -o         Do not add a newline to the end of any received input reports
      -x         Output any received input reports as human readable hex

      For information on specifying a device, see 'hid --help'.

.fam T
.fi
.TP
.B
attached
[\fB-vlp\fP arg]
.PP
.nf
.fam C
      Determine if a matching HID-compliant is attached.

      Wwill output '1' if a matching device is attached, otherwise '0'.

.fam T
.fi
.SH OPTIONS
.TP
.B
\fB--help\fP
Displays help.
.TP
.B
\fB--version\fP
Displays \fBhid\fP tool version information.
.SH EXAMPLE
$ \fBhid\fP list
.SH BUGS
See: https://github.com/coderus-ltd/\fBhid\fP/issues
.SH AUTHOR
Coderus Ltd <info@coderus.com>


hid/common/hid.c
//
//  hid.c
//  hid
//
//  Created by Eddie Woodley on 14/10/2014.
//  Copyright (c) 2014 Coderus. All rights reserved.
//

#include <stdio.h>
#include <string.h>
#include "builtin.h"

#define ARRAY_SIZE(x) (sizeof(x)/sizeof(x[0]))

/// Represents a sub command, its handler and any options it needs.
struct cmd_struct
{
  
  const char *cmd;
  int (*fn)(int, const char **);
  int option;
  
};

/// Define all subcommands and their handlers
static struct cmd_struct commands[] =
{
  { "help", cmd_help, 0 },
  { "version", cmd_version, 0 },
  { "list", cmd_list, 0 },
  { "info", cmd_info, 0 },
  { "getreport", cmd_getreport, 0 },
  { "setreport", cmd_setreport, 0 },
  { "attached", cmd_attached, 0 }
};

/// Run a command from the struct, performing any setup and applying any options.
static int run_builtin(struct cmd_struct *p, int argc, const char **argv)
{
  // Inspect options and perform any setup here.
  // Currently, there are no options.
  return p->fn(argc, argv);
}

int main(int argc, const char * argv[])
{
  int i;
  
  // Adjust argv for convenience
  argv++;
  argc--;
  
  // Find the iteration count
  int iterations = 1;
  for (i = 0; i < argc; i++)
  {
    if(strcmp("-n", argv[i]) == 0 && i+1 < argc)
    {
#pragma warning( push )
#pragma warning( disable : 4996)
        sscanf(argv[i + 1], "%d", &iterations);
#pragma warning( pop ) 
      if(iterations < 1)
      {
        iterations = 1;
      }
      break;
    }
  }
  
  if(argc > 0)
  {
    // Read subcommand
    const char *cmd = argv[0];
    
    // Map --version to version subcommand, adjust argv as needed
    if(strcmp("--version", argv[argc-1]) == 0)
    {
      argv[1] = argv[0];
      argv[0] = cmd = "version";
    }
      
    // Map --help to help subcommand, adjust argv as needed
    if(strcmp("--help", argv[argc-1]) == 0)
    {
      argv[1] = argv[0];
      argv[0] = cmd = "help";
    }
    
    // Find and call the matching subcommand handler
    for (i = 0; i < ARRAY_SIZE(commands); i++)
    {
      struct cmd_struct *p = commands+i;
      if (strcmp(p->cmd, cmd) == 0)
      {
        int ret = 0;
        for(i = 0; i < iterations; i++)
        {
          ret = run_builtin(p, argc, argv);
          if(ret != 0){
            return ret;
          }
        }
        return ret;
      }
    }
    
    // No matching subcommand handler found!
    fprintf(stderr, "hid: '%s' is not a valid command. See 'hid --help'.\n", cmd);
  }
  else
  {
    fprintf(stderr, "hid: subcommand required. See 'hid --help'.\n");
  }
  
  return 255;
  
}
